# BMBL-analysis-notebooks

![LOGO](https://cpb-us-w2.wpmucdn.com/u.osu.edu/dist/0/72768/files/2020/07/bmbl_logo1-300x124.png)

**Warning: This repository is under heavy development and the content is not final yet. If you have early feedback feel free to open issues and get in touch with us.**

## Mission

We are making a series of bioinformatics data analysis notebooks on the current single-cell analysis best practices with data and code examples.

## What this repository is for

The repository is for anyone who wants to:

1. Apply current best-practices analysis using real data.
2. Generate detailed reports to your biological background collaborators.
3. Look for code snippets to quickly generate some results and figures.

## What this repository is NOT

The repository is NOT an introduction to data analysis. There are many courses and resources available, such as [Bioinformatics Training at the Harvard Chan Bioinformatics Core](https://hbctraining.github.io/main/), and [Single-cell best practices from Theis lab](https://www.sc-best-practices.org/preamble.html). Therefore, this repository shies away from these concepts to focus on practical aspects of analysis. To be specific, the repository assumes that users have a basic understanding of the following topics:

1. R, Python, and Unix programming.
2. R markdown settings.
3. Assays such as RNA-seq, single-cell RNA-seq, and Spatial transcriptomics.

## Table of contents

1. Single-cell RNA seq general preprocessing workflow: `/scRNAseq_preprocess_main`
2. A branch of scRNAseq on Immune data analysis: `/scRNAseq_immune_branch`
3. Code Snippets to generate publication level figure: `/figure_code`

In case of questions or problems, please get in touch by posting an issue in this repository or contact me via cankun.wang@osumc.edu

## Acknowledgements

Maintainer: [Cankun Wang](https://github.com/Wang-Cankun)

Contributors:

- [Cankun Wang](https://github.com/Wang-Cankun)
- [Yuzhou Chang](https://github.com/BMEngineeR)