---
title: "Fu_scmulti-omcs"
author: "Qi Guo"
date: '2023-03-13'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# load packages
library(Signac)
library(Seurat)
library(EnsDb.Hsapiens.v86)
library("dplyr")
library(BSgenome.Hsapiens.UCSC.hg38)
dyn.load("/apps/hdf5-serial/gnu/9.1/1.12.0/lib/libhdf5_hl.so.200")
library(hdf5r)
library(GenomicRanges)
library(GenomeInfoDb)
library(qs)
library(sctransform)
set.seed(123)
library(ggplot2)
```

# Integration of eight snATAC-seq

We choose 18-64, which have the largest numbers of features, as the reference sample to ensure that there are common features across the eight datasets since it has the most features. So we can identify as much ATAC features as possible

## 18-64 preprocessing

```{r  include=FALSE}
setwd("/fs/ess/PCON0022/Yuzhou/Shane_ATAC/Primary_data_processing")
counts <- Read10X_h5("18-64_results/filtered_feature_bc_matrix.h5")
fragpath <- "18-64_results/atac_fragments.tsv.gz"

# create a Seurat object containing the RNA adata
multi <- CreateSeuratObject(counts = counts$`Gene Expression`, assay = "RNA")

# Get annotation information
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevelsStyle(annotation) <- "UCSC"

# Create ATAC assay and add it to the object
  multi[["peaks"]] <- CreateChromatinAssay(
    counts = counts$Peaks,
    sep = c(":", "-"),
    fragments = fragpath,
    annotation = annotation
  )
                            
# QC
DefaultAssay(multi) <- "peaks"
multi <- NucleosomeSignal(multi)
multi <- TSSEnrichment(multi)
p <- VlnPlot(
  object = multi,
  features = c(
    "nCount_RNA",
    "nCount_peaks",
    "TSS.enrichment",
    "nucleosome_signal"
    ),
  ncol = 4,
  pt.size = 0
)
setwd("/fs/ess/PCON0022/guoqi/NC-snrna/atac_output")
 ggsave(
   plot = p,
   filename = paste0(names(object)[i], "-qc-vlnplot.png"),
   device = "png",
   dpi = 150,
   width = 12,
   height = 10,
   units = "in"
 )
 
  multi <- subset(
    x = multi,
    subset = nCount_peaks < 40000 &
      nCount_RNA < 75000 &
      nCount_peaks > 1000 &
      nCount_RNA > 1000 &
      nucleosome_signal < 2.5 &
      TSS.enrichment > 1
  )
  p1 <- VlnPlot(
    object = multi,
    features = c(
      "nCount_RNA",
      "nCount_peaks",
      "TSS.enrichment",
      "nucleosome_signal"
    ),
    ncol = 4,
    pt.size = 0
  )

  ggsave(
    plot = p1,
    filename = "18-64_results-qc_ed_-vlnplot.png",
    device = "png",
    dpi = 150,
    width = 12,
    height = 10,
    units = "in"
  )
  # Normalization  
  DefaultAssay(multi) <- "RNA"
  multi <- SCTransform(multi)
  multi <- RunPCA(multi)
  DefaultAssay(multi) <- "peaks"
  multi <- RunTFIDF(multi)
  multi <- FindTopFeatures(multi, min.cutoff = 5)
  multi <- RunSVD(multi)
  qsave(multi, "/fs/ess/PCON0022/guoqi/NC-snrna/atac_output/reference-18_64.qs")
```

## Integrate 18-64 with other seven datasets 

### Preprocess other seven data

```{r  include=FALSE}
setwd("/fs/ess/PCON0022/Yuzhou/Shane_ATAC/Primary_data_processing")
## pre integration
# function
pre_integration <- function(fragpath) {
  fragcounts <- CountFragments(fragments = fragpath)
  atac.cells <- fragcounts[fragcounts$frequency_count > 2000, "CB"]
  # create the fragment object
  atac.frags <-
    CreateFragmentObject(path = fragpath, cells = atac.cells)
  # quantify multiome peaks in the scATAC-seq dataset
  counts <- FeatureMatrix(fragments = atac.frags,
                          features = granges(multi),
                          cells = atac.cells)
  # create object
  atac.assay <- CreateChromatinAssay(counts = counts,
                                     min.features = 1000,
                                     fragments = atac.frags)
  pbmc.atac <-
    CreateSeuratObject(counts = atac.assay, assay = "peaks")
  return(pbmc.atac)
}
# input fragment path, output-target object before filtering and LSI
setwd("/fs/ess/PCON0022/Yuzhou/Shane_ATAC/Primary_data_processing")
target_list<-list()
target_file<-all_file_name[-c(3)]
for(i in 1:length(target_file)){
  fragpath<-paste(target_file[i],"atac_fragments.tsv.gz",sep = "/")
  target_list[[i]]<-pre_integration(fragpath)
}
names(target_list)<-target_file
setwd("/fs/ess/PCON0022/guoqi/NC-snrna/atac_output")
qsave(target_list,"preqc_7samples.qs")#note, first two samples are filtered, so should rerun
#QC
p<-VlnPlot(
  object = target_list[[1]],
  features = c("nCount_peaks", "nFeature_peaks"),
  ncol = 2,
  pt.size = 0
)
target_list[[1]] <- subset(target_list[[1]], nCount_peaks > 1000 & nCount_peaks < 30000)
p<-VlnPlot(
  object = target_list[[2]],
  features = c("nCount_peaks", "nFeature_peaks"),
  ncol = 2,
  pt.size = 0
)
 target_list[[2]] <- subset(target_list[[2]], nCount_peaks > 1000 & nCount_peaks < 30000)

p<-VlnPlot(
  object = target_list[[3]],
  features = c("nCount_peaks", "nFeature_peaks"),
  ncol = 2,
  pt.size = 0
)
target_list[[3]] <- subset(target_list[[3]], nCount_peaks > 1000 & nCount_peaks < 40000)

p<-VlnPlot(
  object = target_list[[4]],
  features = c("nCount_peaks", "nFeature_peaks"),
  ncol = 2,
  pt.size = 0
)
target_list[[4]] <- subset(target_list[[4]], nCount_peaks > 1000 & nCount_peaks < 40000)
p<-VlnPlot(
  object = target_list[[5]],
  features = c("nCount_peaks", "nFeature_peaks"),
  ncol = 2,
  pt.size = 0
)
target_list[[5]] <- subset(target_list[[5]], nCount_peaks > 1000 & nCount_peaks < 30000)

p<-VlnPlot(
  object = target_list[[6]],
  features = c("nCount_peaks", "nFeature_peaks"),
  ncol = 2,
  pt.size = 0
)
target_list[[6]] <- subset(target_list[[6]], nCount_peaks > 1000 & nCount_peaks < 30000)
p<-VlnPlot(
  object = target_list[[7]],
  features = c("nCount_peaks", "nFeature_peaks"),
  ncol = 2,
  pt.size = 0
)
target_list[[7]] <- subset(target_list[[7]], nCount_peaks > 1000 & nCount_peaks < 30000)
```

### Normalization and dimensional reduction

```{r  include=FALSE}
for(i in 1:7){
  target_list[[i]]<-FindTopFeatures(target_list[[i]], min.cutoff = 10)
  target_list[[i]] <- RunTFIDF(target_list[[i]])
  target_list[[i]] <- RunSVD(target_list[[i]])
}
setwd("/fs/ess/PCON0022/guoqi/NC-snrna/atac_output/New_results_qc/")
qsave(target_list,"qc_7samples.qs")
target_list<-qread("qc_7samples.qs")
#annotation
#id
target_list[[1]]$orig.ident<-"1-1"
target_list[[2]]$orig.ident<-"1-7"
target_list[[3]]$orig.ident<-"2-10"
target_list[[4]]$orig.ident<-"2-3"
target_list[[5]]$orig.ident<-"2-5"
target_list[[6]]$orig.ident<-"2-8"
target_list[[7]]$orig.ident<-"T4857"
#stage
target_list[[1]]$stage<-"Control"
target_list[[2]]$stage<-"Late_AD"
target_list[[3]]$stage<-"Late_AD"
target_list[[4]]$stage<-"Mid-AD"
target_list[[5]]$stage<-"Control"
target_list[[6]]$stage<-"Mid-AD"
target_list[[7]]$stage<-"Mid-AD"
#condition
target_list[[1]]$condition<-"Control"
target_list[[2]]$condition<-"AD"
target_list[[3]]$condition<-"AD"
target_list[[4]]$condition<-"AD"
target_list[[5]]$condition<-"Control"
target_list[[6]]$condition<-"AD"
target_list[[7]]$condition<-"AD"
#18-64
multi<-qread("newreference-18_64.qs")
multi$condition<-"Control"
multi$stage<-"Control"
multi$orig.ident<-"18-64"
```

## Integration

```{r  include=FALSE}
ATAC_integration<-function(obj.list,k_weight){
  for(i in 1:length(obj.list)){
    DefaultAssay(obj.list[[i]])<-"peaks"
    obj.list[[i]]<-FindTopFeatures(obj.list[[i]], min.cutoff = 10)
    obj.list[[i]] <- RunTFIDF(obj.list[[i]])
    obj.list[[i]] <- RunSVD(obj.list[[i]])
  }
  integration.anchors <- FindIntegrationAnchors(
    object.list = obj.list,
    anchor.features = rownames(obj.list[["18-64"]]),
    reduction = "rlsi",
    dims = 2:30
  )
  #merge seven samples
  # merge
  if(length(obj.list)==7){
    combined <- merge(obj.list[[1]],c(obj.list[[2]],obj.list[[3]],obj.list[[4]], obj.list[[5]],
                                      obj.list[[6]],obj.list[[7]]))
  }
  else{
    combined <- merge(obj.list[[1]],c(obj.list[[2]],obj.list[[3]],obj.list[[4]], obj.list[[5]],
                                      obj.list[[6]],obj.list[[7]],obj.list[[8]]))
  }
  combined <- FindTopFeatures(combined, min.cutoff = 10)
  combined <- RunTFIDF(combined)
  combined <- RunSVD(combined)
  atac_int <- IntegrateEmbeddings(
    anchorset = integration.anchors,
    reductions = combined[["lsi"]],
    new.reduction.name = "integrated_lsi",
    #dims.to.integrate = 1:30,
    k.weight = k_weight
  )
  atac_int <- RunUMAP(atac_int, reduction = "integrated_lsi", dims = 2:30)
  DefaultAssay(atac_int) <- "peaks"
  atac_int <- FindTopFeatures(atac_int, min.cutoff = 5)
  atac_int <- RunTFIDF(atac_int)
  atac_int <- RunSVD(atac_int)
  return(atac_int)
}

# pbmc.combined<-merge(multi, c(target_list[[1]],target_list[[2]],
#                                target_list[[3]],target_list[[4]],
#                                target_list[[5]],target_list[[6]],target_list[[7]]),
#                       add.cell.ids=c("18_64","1_1","1_7","2_10","2_3","2_5","2_8","T4857"))
# # process the combined dataset
# pbmc.combined <- FindTopFeatures(pbmc.combined, min.cutoff = 10)
# pbmc.combined <- RunTFIDF(pbmc.combined)
# pbmc.combined <- RunSVD(pbmc.combined)
# pbmc.combined <- RunUMAP(pbmc.combined, reduction = "lsi", dims = 2:30)
# p1 <- DimPlot(pbmc.combined, group.by = "orig.ident")
# 
# # find integration anchors
# resul_8<-append(multi,target_list)
# add.cell.ids=c("18_64","1_1","1_7","2_10","2_3","2_5","2_8","T4857")
# 
# for(i in 1:8){
#   resul_8[[i]] <- RenameCells(
#     resul_8[[i]],
#     add.cell.id =add.cell.ids[i]
#   )
# }
# 
# integration.anchors <- FindIntegrationAnchors(
#   object.list = resul_8,
#   anchor.features = rownames(multi),
#   reduction = "rlsi",
#   dims = 2:30
# )
# 
# # integrate LSI embeddings
# 
# 
# integrated <- IntegrateEmbeddings(
#   anchorset = integration.anchors,
#   reductions = pbmc.combined[["lsi"]],
#   new.reduction.name = "integrated_lsi",
#   dims.to.integrate = 1:30
# )
# create a new UMAP using the integrated embeddings
integrated <- RunUMAP(integrated, reduction = "integrated_lsi", dims = 2:30)
p2 <- DimPlot(integrated, group.by = "orig.ident")
qsave(integrated,"atac_8_integration.qs")
```

# Integration of eight snRNA-seq

## QC

```{r  include=FALSE}
setwd("/fs/ess/PCON0022/Yuzhou/Shane_ATAC/Primary_data_processing")
all_file_name <- list.files()
object <- list()
for(i in 1:8){
  a <-
    Read10X_h5(paste0(all_file_name[i], "/filtered_feature_bc_matrix.h5"))
  x <-
    CreateSeuratObject(counts = a[[1]],project =strsplit(all_file_name[i],"_",)[[1]][1] ,min.cells = 3, min.features = 200)
  x[["percent.mt"]] <- PercentageFeatureSet(x, pattern = "^MT-")
  object[i]<-x
}
names(object)<-all_file_name
qsave(object,"rna8objects_filter_preqc.qs")

object_qc<-list()
VlnPlot(object[[1]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
object_qc[[1]] <- subset(object[[1]], subset = nFeature_RNA > 200 & nFeature_RNA < 10000 & percent.mt < 15)
VlnPlot(object[[2]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
object_qc[[2]] <- subset(object[[2]], subset = nFeature_RNA > 200 & nFeature_RNA < 10000 & percent.mt < 15)
VlnPlot(object[[3]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
object_qc[[3]] <- subset(object[[3]], subset = nFeature_RNA > 200 & nFeature_RNA < 12000 & percent.mt < 15)
VlnPlot(object[[4]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
object_qc[[4]] <- subset(object[[4]], subset = nFeature_RNA > 200 & nFeature_RNA < 11000 & percent.mt < 15)
VlnPlot(object[[5]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
object_qc[[5]] <- subset(object[[5]], subset = nFeature_RNA > 200 & nFeature_RNA < 10000 & percent.mt < 15)
VlnPlot(object[[6]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
object_qc[[6]] <- subset(object[[6]], subset = nFeature_RNA > 200 & nFeature_RNA < 10000 & percent.mt < 15)
VlnPlot(object[[7]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
object_qc[[7]] <- subset(object[[7]], subset = nFeature_RNA > 200 & nFeature_RNA < 10000 & percent.mt < 15)
VlnPlot(object[[8]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
object_qc[[8]] <- subset(object[[8]], subset = nFeature_RNA > 200 & nFeature_RNA < 10000 & percent.mt < 15)
```

#annotation

```{r  include=FALSE}
object[[1]]$orig.ident<-"1-1"
object[[2]]$orig.ident<-"1-7"
object[[3]]$orig.ident<-"18-64"
object[[4]]$orig.ident<-"2-10"
object[[5]]$orig.ident<-"2-3"
object[[6]]$orig.ident<-"2-5"
object[[7]]$orig.ident<-"2-8"
object[[8]]$orig.ident<-"T4857"

object[[1]]$stage<-"Control"
object[[2]]$stage<-"Late-AD"
object[[3]]$stage<-"Control"
object[[4]]$stage<-"Late-AD"
object[[5]]$stage<-"Mid-AD"
object[[6]]$stage<-"Control"
object[[7]]$stage<-"Mid-AD"
object[[8]]$stage<-"Mid-AD"

object[[1]]$condition<-"Control"
object[[2]]$condition<-"AD"
object[[3]]$condition<-"Control"
object[[4]]$condition<-"AD"
object[[5]]$condition<-"AD"
object[[6]]$condition<-"Control"
object[[7]]$condition<-"AD"
object[[8]]$condition<-"AD"
```

## Integration

```{r  include=FALSE}
obj.list<-lapply(object_qc,function(x){RenameCells(x,add.cell.id = x$orig.ident[1])})

RNA_integration<-function(obj.list,k_weight){
  obj.list <- lapply(X = obj.list, FUN = SCTransform)
  features <- SelectIntegrationFeatures(object.list = obj.list, nfeatures = 3000)
  obj.list <- PrepSCTIntegration(object.list = obj.list, anchor.features = features)
  anchors <- FindIntegrationAnchors(object.list = obj.list, normalization.method = "SCT",
                                    anchor.features = features)
  combined.sct <- IntegrateData(anchorset = anchors, normalization.method = "SCT", k.weight = k_weight)
  combined.sct <- RunPCA(combined.sct, verbose = FALSE)
  return(combined.sct)
}
```

# Integration of scRNA-seq and scATAC-seq

```{r  include=FALSE}
multi_integration<-function(object_rna,object_atac){
  object_rna[["peaks"]]<-object_atac@assays$peaks
  DefaultAssay(object_rna) <- "peaks"
  object_rna@reductions$lsi<-object_atac@reductions$integrated_lsi
  object_rna@reductions$umap.atac<-object_atac@reductions$umap
  object_rna <- FindMultiModalNeighbors(object_rna, reduction.list = list("pca", "lsi"), dims.list = list(1:50, 2:50))
  object_rna <- RunUMAP(object_rna, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
  return(object_rna)
}
```

# Output

```{r  include=FALSE}
opc_obj_rna<-RNA_integration(opc_obj.list,10)
opc_obj_atac<-ATAC_integration(opc_obj.list,10)
integrate_in<-multi_integration(In_obj_rna,In_obj_atac)
integrate_in<-FindClusters(integrate_in,resolution = 0.5, graph.name = "wsnn", algorithm = 3)
p3 <- DimPlot(integrate_in, reduction = "wnn.umap", group.by = "wsnn_res.0.5", label = TRUE, label.size = 6, repel = TRUE) + ggtitle("WNN_cluster_IN")
p1 <- DimPlot(integrate_in, reduction = "wnn.umap", group.by = "stage", label.size = 6, repel = TRUE) + ggtitle("stage")
p2 <- DimPlot(integrate_in, reduction = "wnn.umap", group.by = "orig.ident", label.size = 6, repel = TRUE) + ggtitle("sampleID")
```