# Exporting seurat object of scRNAseq to shiny cell

This tutorial provides a code procedure for exporting a Seurat object from single-cell RNA sequencing (scRNAseq) to ShinyCell. ShinyCell is a tool that enables interactive exploration and visualization of scRNAseq data. You can view an example of a ShinyCell application at the following link: [https://bmblx.bmi.osumc.edu/scrnaseq_mauro](https://bmblx.bmi.osumc.edu/scrnaseq_mauro)



## Prerequisites:
Before you proceed, ensure you have the following:

A pre-analyzed scRNAseq Seurat object

## Output:
By default, the output object will be saved to the shinyAppMulti folder. Please archive this folder and save it to the OSC folder. Contact Cankun Wang at cankun.wang@osumc.edu to deploy your Shiny app at the BMBL server.

